src/Security/MemberAuthenticator/ChangePasswordHandler.php

```diff
@@ -0,0 +1,309 @@
+        }
+        $token = $request->getVar('t');
+
+        // Check whether we are merely changin password, or resetting.
+        if ($token !== null && $member && $member->validateAutoLoginToken($token)) {
+            $this->setSessionToken($member, $token);
+
+            // Redirect to myself, but without the hash in the URL
+            return $this->redirect($this->link);
+        }
+
+        if (Session::get('AutoLoginHash')) {
+            $message = DBField::create_field(
+                'HTMLFragment',
+                '<p>' . _t(
+                    'SilverStripe\\Security\\Security.ENTERNEWPASSWORD',
+                    'Please enter a new password.'
+                ) . '</p>'
+            );
+
+            // Subsequent request after the "first load with hash" (see previous if clause).
+            return $this->buildResponse($message);
+        }
+
+        if (Security::getCurrentUser()) {
+            // Logged in user requested a password change form.
+            $message = DBField::create_field(
+                'HTMLFragment',
+                '<p>' . _t(
+                    'SilverStripe\\Security\\Security.CHANGEPASSWORDBELOW',
+                    'You can change your password below.'
+                ) . '</p>'
+            );
+
+            return $this->buildResponse($message);
+        }
+        // Show a friendly message saying the login token has expired
+        if ($token !== null && $member && !$member->validateAutoLoginToken($token)) {
+            $customisedController = Controller::curr()->customise(
+                array(
+                    'Content' => DBField::create_field(
+                        'HTMLFragment',
+                        _t(
+                            'SilverStripe\\Security\\Security.NOTERESETLINKINVALID',
+                            '<p>The password reset link is invalid or expired.</p>'
+                            . '<p>You can request a new one <a href="{link1}">here</a> or change your password after'
+                            . ' you <a href="{link2}">logged in</a>.</p>',
+                            [
+                                'link1' => $this->Link('lostpassword'),
+                                'link2' => $this->Link('login')
+                            ]
+                        )
+                    )
+                )
+            );
+
+            return $customisedController->renderWith('changepassword');
+        }
+
+        // Someone attempted to go to changepassword without token or being logged in
+        return Security::permissionFailure(
+            Controller::curr(),
+            _t(
+                'SilverStripe\\Security\\Security.ERRORPASSWORDPERMISSION',
+                'You must be logged in in order to change your password!'
+            )
+        );
+    }
+
+    /**
+     * @param DBField $message
+     * @return DBHTMLText
+     */
+    protected function buildResponse($message)
```

Comment:
I'm not sure if this method naming is right. It's the best I could come up with. Would happily rename if possible.

Also the whole class could use some spring-cleaning, but I think it's fairly decent for now.